django/core/files/storage.py

```diff
@@ -96,6 +95,25 @@ def get_available_name(self, name, max_length=None):
                 name = os.path.join(dir_name, "%s_%s%s" % (file_root, get_random_string(7), file_ext))
         return name
 
+    def generate_filename(self, filename, instance, upload_to):
+        """
+        Creates the final file name or path to be passed to the save method,
+        applying any defined upload_to of the model field and performing file
+        name validation.
+        By default uses the underlying os path operations, but should be overridden
+        to match any other file path manipulation requirement or custom code.
+        """
+        # If upload_to is a callable, make sure that the path it returns is
```

Comment:
This comment seems unnecessary since both paths go through `get_valid_name`(). That being said, some other comments might make the code a bit more readable. It's not obvious why the two branches need to be so different.